crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -269,48 +271,244 @@ impl ClassInfoConstraintFunction {
+                    for right_conj in &from_constraint.disjuncts {
+                        if right_conj.typeguard.is_some() {
+                            // If the right conjunct has a TypeGuard, it "wins" the conjunction
+                            new_disjuncts.push(*right_conj);
+                        } else {
+                            // Intersect the regular constraints
+                            let new_regular = IntersectionBuilder::new(db)
+                                .add_positive(left_conj.constraint)
+                                .add_positive(right_conj.constraint)
+                                .build();
+
+                            new_disjuncts.push(Conjunction {
+                                constraint: new_regular,
+                                typeguard: left_conj.typeguard,
+                            });
+                        }
+                    }
+                }
+
+                *entry.get_mut() = NarrowingConstraint {
+                    disjuncts: new_disjuncts,
+                };
             }
             Entry::Vacant(entry) => {
-                entry.insert(*value);
+                entry.insert(from_constraint.clone());
             }
         }
     }
 }
 
+/// Merge constraints with OR semantics (union/disjunction).
+///
+/// When we have `constraint1 OR constraint2`, we simply concatenate the disjuncts
+/// from both constraints: `(A | B) OR (C | D)` becomes `A | B | C | D`
+///
+/// However, if a place appears in only one branch of the OR, we need to widen it
+/// to `object` in the overall result (because the other branch doesn't constrain it).
 fn merge_constraints_or<'db>(
-    into: &mut NarrowingConstraints<'db>,
-    from: &NarrowingConstraints<'db>,
-    db: &'db dyn Db,
+    into: &mut InternalConstraints<'db>,
+    from: &InternalConstraints<'db>,
+    _db: &'db dyn Db,
 ) {
-    for (key, value) in from {
-        match into.entry(*key) {
+    // For places that appear in `into` but not in `from`, widen to object
+    for (key, value) in &mut into.constraints {
+        if !from.constraints.contains_key(key) {
+            *value = NarrowingConstraint::regular(Type::object());
+        }
+    }
+
+    for (key, from_constraint) in &from.constraints {
+        match into.constraints.entry(*key) {
             Entry::Occupied(mut entry) => {
-                *entry.get_mut() = UnionBuilder::new(db).add(*entry.get()).add(*value).build();
+                // Simply concatenate the disjuncts
+                entry
+                    .get_mut()
+                    .disjuncts
+                    .extend(from_constraint.disjuncts.clone());
             }
             Entry::Vacant(entry) => {
-                entry.insert(Type::object());
+                // Place only appears in `from`, not in `into`.
+                // Widen to object since the other branch doesn't constrain it.
+                entry.insert(NarrowingConstraint::regular(Type::object()));
```

Comment:
I wonder why we don't just do nothing here? We don't want to constrain, and there's already no constraint in `into`, so let's just... leave it that way. Seems to pass tests.
```suggestion
                // Place only appears in `from`, not in `into`. No constraint needed.
```